backend/regalloc/regalloc_split_state.ml

```diff
@@ -106,6 +107,139 @@ end = struct
     !destructions_at_end, definitions_at_beginning
 end
 
+(* Optimizes `destructions_at_end` by avoiding the repeated spill of a constant.
+   If a register is set only once, then all spills of that register dominated by
+   a spill of the same register can be deleted. *)
+module RemoveDominatedSpillsForConstants : sig
+  val optimize :
+    Cfg_with_liveness.t ->
+    Cfg_dominators.t ->
+    destructions_at_end:destructions_at_end ->
+    destructions_at_end
+end = struct
+  (* CR-someday xclerc for xclerc: move to Cfg_loop_infos? *)
+  let is_in_loop : Cfg_loop_infos.t -> Label.t -> bool =
+   fun loops label ->
+    Cfg_loop_infos.EdgeMap.exists
+      (fun _ (loop : Cfg_loop_infos.loop) -> Label.Set.mem label loop)
+      loops.loops
+
+  type set =
+    | Exactly_once
+    | At_least_twice
+
+  let string_of_set = function
+    | Exactly_once -> "exactly once"
+    | At_least_twice -> "at least twice"
+
+  let rec remove_dominated_spills :
+      Cfg_dominators.t ->
+      Cfg_loop_infos.t ->
+      Cfg_dominators.dominator_tree ->
+      num_sets:set Reg.Tbl.t ->
+      already_spilled:Label.t Reg.Map.t ->
+      destructions_at_end:destructions_at_end ->
+      destructions_at_end =
+   fun doms loops tree ~num_sets ~already_spilled ~destructions_at_end ->
+    if split_debug then log ~indent:1 "remove_dominated_spills %d" tree.label;
+    let already_spilled = ref already_spilled in
+    let destructions_at_end : (destruction_kind * Reg.Set.t) Label.Map.t =
+      Label.Map.update tree.label
+        (function
+          | Some (Destruction_on_all_paths, destroyed) ->
+            let destroyed : Reg.Set.t =
+              Reg.Set.filter
+                (fun (reg : Reg.t) ->
+                  if split_debug
+                  then log ~indent:2 "register %a" Printmach.reg reg;
+                  let keep =
+                    match Reg.Tbl.find_opt num_sets reg with
+                    | None | Some At_least_twice -> true
```

Comment:
How is `None` case reachable here?
```suggestion
                    | None  -> assert false
                    | Some At_least_twice -> true
```

Reply:
It is essentially defensive. Reaching this case
would mean spilling a value never set. While
it cannot happen now -I think-, it is not totally
crazy: a value set "by the calling conventions"
may be spilled and yet never set in the function.

(The reason why it should not happen right now
is because we spill only values whose location
is unknown and the current calling conventions
would use either an hardware register or the
domain.)

Reply:
ok, I'd argue for an assertion here and I think it would it trip the validator, wouldn't it?

Reply:
I am not sure I follow: we currently "receive"
parameter values in `rax` and co.